File: app/src/main/java/de/jerleo/database/TableMeter.java
package de.jerleo.database;

import android.content.ContentValues;
import android.database.Cursor;

import java.util.ArrayList;
import java.util.List;

import de.jerleo.database.DatabaseColumn.DataType;
import de.jerleo.model.Meter;

import static android.provider.BaseColumns._ID;
import static de.jerleo.database.Constants.METER;
import static de.jerleo.database.Constants.NAME;
import static de.jerleo.database.Constants.NUMBER;
import static de.jerleo.database.Constants.PRIOR;
import static de.jerleo.database.Constants.UNIT;

public class TableMeter extends DatabaseTable {

    public TableMeter(Database database) {

        super(database, METER, true);

        addColumn(new DatabaseColumn(NUMBER, DataType.TEXT));
        addColumn(new DatabaseColumn(NAME, DataType.TEXT));
        addColumn(new DatabaseColumn(UNIT, DataType.TEXT));
        addColumn(new DatabaseColumn(PRIOR, DataType.INTEGER));

        final DatabaseColumn meter = getColumn(PRIOR);
        meter.setForeignKeyConstraint(true);
        meter.setForeignTable(METER);
        meter.setForeignColumn(_ID);
    }

    @Override
    public ContentValues getValues(Object object, boolean forUpdate) {

        final Meter meter = (Meter) object;

        final ContentValues values = new ContentValues();
        if (forUpdate)
            values.put(_ID, meter.getId());
        values.put(NUMBER, meter.getNumber());
        values.put(NAME, meter.getName());
        values.put(UNIT, meter.getUnit().name());
        values.put(PRIOR, meter.hasPrior() ? meter.getPrior().getId() : null);

        return values;
    }

    @Override
    public String getWhereClause(Object object) {

        final Meter meter = (Meter) object;
        return _ID + " = " + meter.getId();
    }

    @Override
    public List<Meter> read(String condition) {

        final ArrayList<Meter> meters = new ArrayList<>();
        final String[] columns = {_ID, NUMBER, NAME, UNIT, PRIOR};
        final Cursor cursor = database.query(METER, columns, condition, null,
                null, null, null);
        while (cursor.moveToNext()) {
            final Meter meter = new Meter();
            int col = 0;
            meter.setId(cursor.getLong(col++));
            meter.setNumber(cursor.getString(col++));
            meter.setName(cursor.getString(col++));
            meter.setUnit(Meter.Unit.valueOf(cursor.getString(col++)));
            meter.setPrior(cursor.getLong(col));
            meters.add(meter);
        }
        return meters;
    }

}

File: app/src/main/kotlin/de/jerleo/database/table/TableMeter.kt
package de.jerleo.database.table

import android.content.ContentValues
import android.provider.BaseColumns
import de.jerleo.database.Column
import de.jerleo.database.Constants
import de.jerleo.database.Database
import de.jerleo.database.Table
import de.jerleo.model.Meter

class TableMeter(database: Database) :
    Table(database, Constants.METER, true) {

    init {
        add(Column(Constants.NUMBER, Column.Type.TEXT))
        add(Column(Constants.NAME, Column.Type.TEXT))
        add(Column(Constants.UNIT, Column.Type.TEXT))
        add(Column(Constants.PRIOR, Column.Type.INTEGER).apply {
            isForeignKeyConstraint = true
            foreignTable = Constants.METER
            foreignColumn = BaseColumns._ID
        })
    }

    override fun values(obj: Any, forUpdate: Boolean): ContentValues {
        val meter = obj as Meter
        val values = ContentValues()
        if (forUpdate) values.put(BaseColumns._ID, meter.id)
        values.put(Constants.NUMBER, meter.number)
        values.put(Constants.NAME, meter.name)
        values.put(Constants.UNIT, meter.unit.name)
        meter.prior?.let { values.put(Constants.PRIOR, meter.prior!!.id) }
        return values
    }

    override fun whereClause(obj: Any): String = BaseColumns._ID + " = " + (obj as Meter).id

    override fun read(condition: String?): MutableList<Meter> {
        val meters = ArrayList<Meter>()
        val columns = arrayOf(
            BaseColumns._ID,
            Constants.NUMBER,
            Constants.NAME,
            Constants.UNIT,
            Constants.PRIOR
        )
        val cursor = database.query(
            Constants.METER, columns, condition, null,
            null, null, null
        )
        while (cursor.moveToNext()) {
            val meter = Meter()
            var col = 0
            meter.id = cursor.getLong(col++)
            meter.number = cursor.getString(col++)
            meter.name = cursor.getString(col++)
            meter.unit = Meter.Unit.valueOf(cursor.getString(col++))
            meter.priorId = cursor.getLong(col)
            meters.add(meter)
        }
        return meters
    }

    fun import(): MutableList<Meter> {
        val meters: MutableList<Meter> = read(null)
        for (meter in meters) {
            val id = meter.id.toString()
            val tariffs = Database.tableTariff.read(id)
            val readings = Database.tableReading.read(id)
            meter.tariffs = tariffs
            meter.readings = readings
            meter.update()
        }
        return meters
    }
}